pkg/controllers/jobset_controller.go

```diff
@@ -327,13 +328,39 @@ func (r *JobSetReconciler) resumeJobSetIfNecessary(ctx context.Context, js *jobs
 	log := ctrl.LoggerFrom(ctx)
 
 	nodeAffinities := map[string]map[string]string{}
+	replicatedJobStatus := r.calculateReplicatedJobStatuses(ctx, js, ownedJobs)
 	for _, replicatedJob := range js.Spec.ReplicatedJobs {
 		nodeAffinities[replicatedJob.Name] = replicatedJob.Template.Spec.Template.Spec.NodeSelector
 	}
 
+	checkIfJobShouldbeResumed := func(js *jobset.JobSet, replicatedJobStatus []jobset.ReplicatedJobStatus, replicatedJobName string) startupPolicyAction {
+		if inOrderStartupPolicy(js.Spec.StartupPolicy) {
+			// If startup policy is enabled,
+			// we are checking against replicated job status.
+			// StartupPolicy will resume jobs in order.
+			for _, rjobStatus := range replicatedJobStatus {
+				if replicatedJobName == rjobStatus.Name {
+					log.V(2).Info("testing startup policy against replicated job status", "ReplicatedJobStatus", rjobStatus)
+					return matchStartupPolicy(js, rjobStatus)
+				}
+			}
+		}
+		return resume
+
+	}
+
 	// If JobSpec is unsuspended, ensure all active child Jobs are also
 	// unsuspended and update the suspend condition to true.
 	for _, job := range ownedJobs.active {
+		replicatedJobName := job.Labels[jobset.ReplicatedJobNameKey]
```

Comment:
As discussed in https://github.com/kubernetes-sigs/jobset/pull/246#discussion_r1452760904, instead of iterating over active jobs, we need to iterate over the replicatedJobs in order just like we are doing in createJobs. 

To do this, create a map of replicatedJobName to active jobs, and iterate over the replicatedJobs just like we did in the createJobs function and lookup the active jobs to unsuspend from the map we created above.

```
replicatedJobs map[string][]*batchv1.Job
// populate replicatedJobs map
for _, rjob := range js.Spec.ReplicatedJobs {
  if unsuspend replicatedJob {
     for job := range replicatedJobs[rjob.Name] {
            .....
     }
  }
}
```

Reply:
Thanks for the pointer. I applied this change and I did a e2e test with both suspend and creation. They both are working now!